util/base_parser.h

```diff
@@ -42,15 +52,32 @@ class BaseParser {
     }
 
     constexpr void skip_whitespace() {
-        while (!is_eof() && is_space(peek())) {
+        while (!is_eof() && is_whitespace(peek())) {
             advance(1);
         }
     }
 
-private:
-    static constexpr auto space_chars = std::array{' ', '\f', '\n', '\r', '\t', '\v'};
-    constexpr static bool is_space(char c) { return std::ranges::find(space_chars, c) != end(space_chars); }
+    static constexpr bool is_whitespace(char c) {
+        return c == ' ' || c == '\f' || c == '\n' || c == '\r' || c == '\t' || c == '\v';
+    }
+
+    static constexpr bool is_c0(char c) { return c >= 0x00 && c <= 0x1f; }
+
+    static constexpr bool is_c0_or_space(char c) { return is_c0(c) || c == ' '; }
 
+    static constexpr bool is_tab_or_newline(char c) { return c == '\t' || c == '\n' || c == '\r'; }
+
+    static constexpr bool is_upper_alpha(char c) { return c >= 'A' && c <= 'Z'; }
+
+    static constexpr bool is_lower_alpha(char c) { return c >= 'a' && c <= 'z'; }
+
+    static constexpr bool is_digit(char c) { return c >= '0' && c <= '9'; }
+
+    static constexpr bool is_alpha(char c) { return is_upper_alpha(c) || is_lower_alpha(c); }
+
+    static constexpr bool is_alphanumeric(char c) { return is_alpha(c) || is_digit(c); }
```

Comment:
Some of these methods are already available in `util/string.h`.

Reply:
I moved them to string.h, and removed the duplicate definition of is_whitespace() from BaseParser.